atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java

```diff
@@ -91,6 +122,77 @@ public void instrumentWithCustomNameAll() {
         assertMethodInstrumented(MetricRegistry.name(CUSTOM_METRIC_NAME, PING_NOT_TIMED_METHOD), service::pingNotTimed);
     }
 
+    @Test
+    public void instrumentTaggedAsyncFunction() throws InterruptedException, ExecutionException {
+        AsyncTestService asyncTestService = AtlasDbMetrics.instrumentWithTaggedMetrics(taggedMetrics,
+                AsyncTestService.class, this.asyncTestService);
+
+        String asyncPingMetricName = MetricRegistry.name(AsyncTestService.class, "asyncPing");
+        assertTimerNotRegistered(asyncPingMetricName);
+
+        List<ListenableFuture<String>> futures = IntStream.range(0, 10)
+                .mapToObj($ -> asyncTestService.asyncPing())
+                .collect(Collectors.toList());
+
+        Instant now = Instant.now();
+        MetricName metricName = MetricName.builder().safeName(asyncPingMetricName).build();
+        ListenableFuture<Boolean> done = Futures.whenAllSucceed(futures).call(() -> {
+            // have to do it this because we can't edit the future we get back and it's only a callback as opposed to a
+            // transformed future
+            Awaitility.await()
+                    .atMost(ASYNC_DURATION_TTL.toMillis(), TimeUnit.MILLISECONDS)
+                    .until(() -> taggedMetrics.timer(metricName).getSnapshot().size() > 0);
+            return true;
+        }, MoreExecutors.directExecutor());
+
+        assertThat(done.get()).isEqualTo(true);
+
+        assertThat(done.get()).isEqualTo(true);
+        assertThat(Instant.now())
+                .as("in the event of scheduling issues, and despite having 10 concurrent futures, we complete "
+                        + "everything with 2*ttl")
+                .isBefore(now.plus(ASYNC_DURATION_TTL.plus(ASYNC_DURATION_TTL)));
+        Snapshot snapshot = taggedMetrics.timer(metricName).getSnapshot();
+        assertThat(Duration.ofNanos(snapshot.getMin())).isGreaterThan(ASYNC_DURATION_TTL);
+        assertThat(snapshot.size()).isEqualTo(10);
+    }
+
+    @Test
+    public void instrumentTaggedAsyncFunctionWithExtraTags() throws InterruptedException, ExecutionException {
```

Comment:
Copied code from above - can a common method be factored out?

Reply:
some code can be factored out, I won't go the full length because it's important to see what's actually in the test without having to jump around everywhere.

Reply:
That's fair, it just seemed that a fair amount was exact copy/paste (but I don't mind things with small variations not being factored out exactly).